src/Bounds.cpp

```diff
@@ -1235,8 +1235,27 @@ class Bounds : public IRVisitor {
                                        !b_interval.min.type().is_uint() &&
                                        can_prove(b_interval.min < 0 &&
                                                  b_interval.min > -t.bits())) {
-                                interval.min = a_interval.min >> abs(b_interval.min);
+                                if (a_interval.min.type().is_uint() || can_prove(a_interval.min >= 0)) {
+                                    interval.min = a_interval.min >> abs(b_interval.min);
+                                } else if (can_prove(a_interval.min < 0)) {
+                                    if (b_interval.has_upper_bound()) {
+                                        if (can_prove(b_interval.max <= 0)) {
+                                            // If b is strictly non-positive, then the magnitude can only decrease.
+                                            interval.min = a_interval.min;
+                                        } else {
+                                            // If b could be positive, then the magnitude might increase.
+                                            interval.min = min(a_interval.min, a_interval.min << b_interval.max);
+                                        }
+                                    }
+                                }
+                                // TODO: Are there any other cases we can handle here?
+                            } else if (a_interval.has_lower_bound() &&
+                                    b_interval.has_upper_bound() &&
+                                    can_prove(a_interval.min >= 0 && b_interval.max <= 0)) {
```

Comment:
Why does this require b_interval.max <= 0?

Reply:
Ah, because I mistakenly thought that it did, and z3 didn't complain that there was an extra + unnecessary condition. Just verified that this works without `b` being bounded.

Reply:
Fixed in c4b7b8a